User-as-a-case

kaapstorm opened the pull request:
This feature allows user data and activity to be tracked in a case.

Case properties are available using the prefix "user:", e.g. "user:dob" or "user:last_activity". The user case generally behaves like any other case.

kaapstorm committed:
Hid Call Center "case type" and defaulted to "user-case"

Also added README.rst. Will extend Implementation section as work
progresses.

kaapstorm committed:
Add "user_id" to user-case

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Allow colons in property names

kaapstorm committed:
Add usercase node to xform

kaapstorm committed:
str(xform) returns XML

kaapstorm committed:
Add bind and setvalue nodes to xform for usercase

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/xform.py

kaapstorm committed:
Add usercase to advanced XForm

kaapstorm committed:
Add usercase_enabled to template context

Also, reuse get_details function and clean some imports

kaapstorm committed:
When form has no case, show usercase management

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Include user case properties in get_all_case_properties()

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Only bind usercase if necessary

kaapstorm committed:
Fixed bug: case_type is a string, not a set.

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Removed GUI changes for now

kaapstorm committed:
WIP: Add usercase_id to session.

Test normal form

TODO: Test advanced form, use xpath

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/suite_xml.py
	corehq/apps/app_manager/tests/test_suite.py

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Use XPath

dimagimon commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -323,6 +323,37 @@ def test_tiered_select_with_advanced_module_as_parent(self):
+# Just use AdvancedModule
+
+
+    def test_usercase_id_added_update(self):
```
E303 too many blank lines (2)

dimagimon commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +980,93 @@ def already_has_meta(self):
+        usercase_bound = False
+
+        if 'update_case' in actions and hasattr(actions['update_case'], 'update'):
+            usercase_updates = {k[5:]: v for k, v in actions['update_case'].update.items() if k.startswith('user:')}
```
E501 line too long (116 > 115 characters)

benrudolph replied:
I'm assuming the `5:` is cutting off the 'user:' prefix. Small, but might be nice if we had a constant for `"user:"`. Then we could reduce spelling errors ever and in this scenario we could do `k[len(USERCASE_PREFIX):]` and it'd be clear that we're chopping off the user prefix

dimagimon commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +980,93 @@ def already_has_meta(self):
+
+        if 'case_preload' in actions:
+            self.add_casedb()
+            usercase_preloads = {k: v[5:] for k, v in actions['case_preload'].preload.items() if v.startswith('user:')}
```
E501 line too long (119 > 115 characters)

dimagimon commented on corehq/apps/app_manager/xpath.py:
```diff
@@ -12,8 +12,8 @@ def dot_interpolate(string, replacement):
-def session_var(var, subref=None):
-    return XPath(u"instance('commcaresession')/session/{0}data/{1}".format(subref + '/' if subref else '', var))
+def session_var(var, subref=None, data=u'data'):
+    return XPath(u"instance('commcaresession')/session/{0}{1}/{2}".format(subref + '/' if subref else '', data, var))
```
E501 line too long (117 > 115 characters)

kaapstorm commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -323,6 +323,37 @@ def test_tiered_select_with_advanced_module_as_parent(self):
 
         self.assertXmlPartialEqual(self.get_xml('advanced_module_parent'), app.create_suite(), "./entry[1]")
 
+# Just use AdvancedModule
```
Tests for updating from and preloading into advanced forms pending.

kaapstorm committed:
PEP8

benrudolph commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1581,6 +1604,14 @@ def get_datum_session_var(action, action_index):
 
+        def uses_usercase(form):
+            for action in form.actions.load_update_cases:
+                if action.preload:
```
could this be simplified to 
```
if action.preload or action.case_properties
  return any(k.startswith('user:') for k in action.preload.keys())
```

kaapstorm replied:
:+1:

benrudolph commented on corehq/apps/app_manager/tests/test_suite.py:
```diff
@@ -393,7 +424,7 @@ def test_case_list_registration_form_advanced(self):
         update_form = app.new_form(1, 'Update Case', lang='en')
         update_form.unique_id = 'update_case_form'
-        update_form.actions.load_update_cases.append(LoadUpdateAction(
+        update_form.actions.load_update_cases.append(LoadUpdateAction(  # <-- e.g.
```
probably don't need this anymore

kaapstorm replied:
True. Currently working on two tests for advanced forms and that comment is gone. (If I'd noticed that comment earlier, I wouldn't have committed it.)

benrudolph commented on corehq/apps/app_manager/views.py:
```diff
@@ -867,33 +912,6 @@ def case_list_form_options(case_type):
         }
     elif isinstance(module, AdvancedModule):
         case_type = module.case_type
-        def get_details():
```
+1 for pulling this out

benrudolph commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +980,93 @@ def already_has_meta(self):
+                    )
+
+            if action.case_properties:
+                usercase_updates = {k[5:]: v for k, v in action.case_properties.items() if k.startswith('user:')}
```
this is coming up a lot, maybe a mall utility function could be useful

benrudolph commented:
cool lookin nice, excited for this

kaapstorm committed:
Created a few utility functions.

Also created const.USERCASE_PREFIX and cleaned up a little.

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1581,6 +1605,12 @@ def get_datum_session_var(action, action_index):
+        def uses_usercase(form):
+            for action in form.actions.load_update_cases:
+                if action.preload or action.case_properties:
+                    return any_usercase_items(action.preload.keys())
```
don't you also need to check `action.case_properties`?

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1407,11 +1416,26 @@ def case_sharing_requires_assertion(form):
+                return any_usercase_items(actions['update_case'].update.keys())
+            return False
+
+        def preloads_usercase(form):
```
Seems like you could combine these two functions since they only ever get used together

snopoke commented on corehq/apps/app_manager/util.py:
```diff
@@ -146,11 +149,24 @@ def get_properties(self, case_type, already_visited=(),
+        #           parent case type's properties will be returned as
+        #           `parent/user:property`. That's probably unlikely, and a
+        #           good thing that it is, because that looks like a mess, and
+        #           will likely cause problems.
```
This may not be as unlikely as you think. I can quite easily foresee people creating child cases of the user case e.g. one case per exam.

I think its fine not to handle that now but is likely something we'll need to figure out.

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +981,93 @@ def already_has_meta(self):
+        actions = form.active_actions()
+        usercase_bound = False
+
+        if 'update_case' in actions and hasattr(actions['update_case'], 'update'):
```
Why is the `hasattr` check needed here?

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +981,93 @@ def already_has_meta(self):
+            self.add_casedb()
+            usercase_preloads = get_usercase_values(actions['case_preload'].preload.items())
+            if usercase_preloads and not usercase_bound:
+                self._add_usercase_bind(usercase_path)
```
You don't need to add a bind for preloading, only for updating.

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +981,93 @@ def already_has_meta(self):
+                self._add_usercase_bind(usercase_path)
+                usercase_bound = True
+            for nodeset, property_ in usercase_preloads.items():
+                parent_path, property_ = split_path(property_)
```
I don't think this is necessary since the user case will never have a parent.

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -970,6 +981,93 @@ def already_has_meta(self):
+                    value=id_xpath.case().property(property_xpath),
+                )
+
+    def add_usercase_advanced(self, form):
```
I'm thinking we should handle the advanced forms differently and rather than mixing in the user case properties in other actions we should just have a separate action with only user case properties. That way we wouldn't have to add any new code to suite_xml.py or xform.py to for advanced forms / modules.

Can chat more about this offline.

snopoke commented on corehq/apps/app_manager/xpath.py:
```diff
@@ -112,6 +113,10 @@ def case(self):
 
 
+class UserCaseXPath(CaseSelectionXPath):
+    selector = 'user_id'
```
I think this should be more specific:
```
instance('casedb')/casedb/case[@case_type='user_case'][hq_user_id={user_id in session}]
```
See also: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/callcenter/utils.py#L73-73

snopoke commented on corehq/apps/callcenter/utils.py:
```diff
@@ -32,6 +32,7 @@ def valid_element_name(name):
     fields.update({
         'name': commcare_user.name or commcare_user.raw_username,
         'username': commcare_user.raw_username,
+        'user_id': commcare_user.get_id,
```
this already gets added: https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/callcenter/utils.py#L73-73

kaapstorm committed:
Merge branch 'master' into usercase

snopoke commented:
bump

kaapstorm committed:
Combine updates_usercase and preloads_usercase

kaapstorm committed:
Advanced module can already support multiple cases.

No need to add special functionality just for user cases.
Also removed binds for preloads.

kaapstorm committed:
Pass user case type to Suite Generator

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py
	corehq/apps/app_manager/views.py

kaapstorm committed:
Merge branch 'usercase' of github.com:dimagi/commcare-hq into usercase

Conflicts:
	corehq/apps/app_manager/suite_xml.py
	corehq/apps/app_manager/util.py

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -18,14 +18,16 @@
     SuiteValidationError,
 )
 from corehq.apps.app_manager import id_strings
+from corehq.apps.app_manager.const import CAREPLAN_GOAL, CAREPLAN_TASK, SCHEDULE_LAST_VISIT, SCHEDULE_PHASE, \
```
F401 'CASE_ID' imported but unused

dimagimon replied:
F401 'USERCASE_PREFIX' imported but unused

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1664,6 +1694,9 @@ def get_datum_session_var(action, action_index):
         if self.app.case_sharing and case_sharing_requires_assertion(form):
             self.add_case_sharing_assertion(e)
 
+        if uses_usercase(form):
```
F821 undefined name 'uses_usercase'

dannyroberts commented:
:X:

czue commented:
```
3.49s$ coverage run manage.py test --noinput --failfast --traceback --testrunner=$TEST_RUNNER
Traceback (most recent call last):
  File "manage.py", line 65, in <module>
    execute_from_command_line(sys.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 453, in execute_from_command_line
    utility.execute()
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 392, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 272, in fetch_command
    klass = load_command_class(app_name, subcommand)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 77, in load_command_class
    module = import_module('%s.management.commands.%s' % (app_name, name))
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/utils/importlib.py", line 35, in import_module
    __import__(name)
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/south/management/commands/__init__.py", line 10, in <module>
    import django.template.loaders.app_directories
  File "/home/travis/virtualenv/local/lib/python2.7/site-packages/django/template/loaders/app_directories.py", line 25, in <module>
    raise ImproperlyConfigured('ImportError %s: %s' % (app, e.args[0]))
ImproperlyConfigured: ImportError corehq.apps.userhack: cannot import name XForm
```

snopoke commented:
That usually means the is a circular import reference somewhere.

kaapstorm commented:
Yes. :-( And missing `uses_usercase` was accidentally deleted in a bad merge.

kaapstorm committed:
Fixed bad merge

Adopted consistent spelling for `usercase_type`.

Circular import still broken.

kaapstorm committed:
Fixed circular reference

Tests still broken

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Fixed failing tests

Updated function for selecting usercase_id in form XML.

kaapstorm committed:
Removed unnecessary field

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/views.py

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/suite_xml.py

kaapstorm committed:
Assign user case type to a constant

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

kaapstorm committed:
Use dict.get() correctly.

kaapstorm committed:
Use get_extra_case_id_datums for user case

get_extra_case_id_datums adds user case ID to session. Also used by SMS forms
and CloudCare.

kaapstorm committed:
CloudCare and SMSForms to get real user case ID

Previously we were passing them a new ID for the user case each time a form
that uses the user case was opened.

kaapstorm committed:
Mock get_usercase_type for Form Linking tests

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
get_case_by_domain_hq_user_id returns None

It doesn't raise NoResultFound

dimagimon commented on corehq/apps/cloudcare/views.py:
```diff
@@ -1,11 +1,14 @@
-from couchdbkit import ResourceConflict
+from corehq.apps.app_manager.const import USERCASE_ID
+from corehq.apps.hqcase.utils import get_case_by_domain_hq_user_id
+from couchdbkit import ResourceConflict, NoResultFound
```
F401 'NoResultFound' imported but unused

dimagimon commented on corehq/apps/smsforms/app.py:
```diff
@@ -1,5 +1,8 @@
+from corehq.apps.app_manager.const import USERCASE_ID
 from corehq.apps.app_manager.suite_xml import SuiteGenerator
+from corehq.apps.hqcase.utils import get_case_by_domain_hq_user_id
+from couchdbkit import NoResultFound
```
F401 'NoResultFound' imported but unused

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Unused import

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/tests/test_suite.py

dimagimon commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -21,14 +21,16 @@
 )
 from corehq.toggles import MODULE_FILTER
 from corehq.apps.app_manager import id_strings
+from corehq.apps.app_manager.const import CAREPLAN_GOAL, CAREPLAN_TASK, SCHEDULE_LAST_VISIT, SCHEDULE_PHASE, \
```
F401 'CASE_ID' imported but unused

kaapstorm committed:
Use constant user case type "commcare-user"

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/views.py

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Separate call center from usercase

dannyroberts committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/xform.py

kaapstorm committed:
create_user_cases should be a Celery task

kaapstorm committed:
Rename functions to singular

And don't copy custom user data to usercase

kaapstorm committed:
Return None if domain_name not found

kaapstorm committed:
Reuse any_usercase_items function

kaapstorm committed:
Use strict=True fetching domain to enable usercase

kaapstorm committed:
Merge branch 'master' into usercase

Conflicts:
	corehq/apps/app_manager/views.py
	corehq/apps/app_manager/xform.py

kaapstorm committed:
Merge branch 'usercase' of github.com:dimagi/commcare-hq into usercase

Conflicts:
	corehq/apps/app_manager/xform.py

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Handle when domain_name is None

kaapstorm committed:
Lock enable_usercase with CriticalSection

kaapstorm committed:
Lock sync_user_case with CriticalSection

dimagimon commented on corehq/apps/callcenter/utils.py:
```diff
@@ -3,87 +3,111 @@
+                return False
+
+        # remove any keys that aren't valid XML element names
+        fields = {k: v for k, v in commcare_user.user_data.items() if valid_element_name(k)} if copy_user_data else {}
```
E501 line too long (118 > 115 characters)

snopoke replied:
Still need to remove these from the user case (but keep them for the call center cases)

kaapstorm commented:
Thanks for all the feedback, @snopoke, @gcapalbo. Updates pushed.

I've got some questions about CriticalSection. We only seem to use it in these places:
* corehq/apps/reminders/models.py
* corehq/apps/reminders/views.py
* corehq/apps/reminders/tasks.py
* corehq/apps/reminders/event_handlers.py
* corehq/apps/sms/verify.py
* corehq/apps/app_manager/util.py
* corehq/apps/kookoo/views.py
* submodules/dimagi-utils-src/dimagi/utils/multithreading.py
* submodules/dimagi-utils-src/dimagi/utils/couch/\_\_init\_\_.py

We don't seem to use it for syncing cases, even user cases for call center. How come? Is that potentially a bug in call center? 

I've added it in both places you suggested, @gcapalbo, and I've committed them separately in case I need to revert.

[The lock in enable_usercase][1] will return quickly because creating the usercases happens asynchronously, and if a process is waiting for the lock, when it's released they'll find that usercase is already enabled on the domain and that there is nothing to do. So that seems right. I think this will only get called because of a change made in the app builder. Or am I missing another scenario? What are the chances that two users will be modifying the same app simultaneously? Even though the chances are small, should we lock it on principle, or not bother?

Regarding [locking in sync_user_case][2], we're using a different key to the one we use in enable_usercase because we're locking on each usercase, instead of all usercases for a domain. What happens if an FM enables usercase at the same time that a user is saved? That user's case will be synced with two different locks. Is this something we need to consider? The chances seem really small. And if there is a race condition, what would break?


  [1]: https://github.com/dimagi/commcare-hq/commit/21d3b863f384d92359de00725cea5917061830a9
  [2]: https://github.com/dimagi/commcare-hq/commit/5496d31acbb2bc09be35e66b442796f0de940e14

snopoke commented:
That all seems fine. I think the only scenario we should care about in the `sync_user_case` function is if two cases get created for the same user but the lock should prevent that.

snopoke commented on corehq/apps/app_manager/views.py:
```diff
@@ -1060,6 +1066,7 @@ def qualified_form_name(form):
     })
 
     context['latest_commcare_version'] = get_commcare_versions(request.user)[-1]
+    context['usercase_enabled'] = Domain.get_by_name(domain).call_center_config.enabled
```
Need to change this to use new domain property.

snopoke commented on corehq/apps/cloudcare/views.py:
```diff
@@ -199,6 +202,9 @@ def form_context(request, domain, app_id, module_id, form_id):
     session_extras.update({datum['datum'].id: uuid.uuid4().hex for datum in datums})
+    usercase = get_case_by_domain_hq_user_id(domain, request.couch_user.get_id, include_docs=False)
+    if usercase:
+        session_extras[USERCASE_ID] = usercase['id']
```
Should only add this if required.

snopoke commented on corehq/apps/domain/forms.py:
```diff
@@ -392,9 +393,10 @@ class DomainGlobalSettingsForm(forms.Form):
-        help_text=_("Enter the case type to be used for FLWs in call center apps")
+        # Hide the case type, but keep the form field in order to preserve
+        # existing values on update.
+        widget=forms.HiddenInput()
```
Can undo this change.

snopoke commented on corehq/apps/smsforms/app.py:
```diff
@@ -59,6 +61,10 @@ def start_session(domain, contact, app, module, form, case_id=None, yield_respon
+        if contact.doc_type == 'CommCareUser':
+            usercase = get_case_by_domain_hq_user_id(domain, contact.get_id, include_docs=False)
+            if usercase:
+                session_data[USERCASE_ID] = usercase['id']
```
again, only do this if necessary

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1490,6 +1496,28 @@ def get_auto_select_assertions(self, case_id_xpath, mode, locale_arguments=None)
+            return False
+
+        datums = []
+        if form.form_type == 'module_form' and uses_usercase(form) and self.is_usercase_enabled:
```
if the form has 'user case' properties but `self.is_usercase_enabled` is false that seems like an error that should prevent the build.

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1490,6 +1496,28 @@ def get_auto_select_assertions(self, case_id_xpath, mode, locale_arguments=None)
+        datums = []
+        if form.form_type == 'module_form' and uses_usercase(form) and self.is_usercase_enabled:
+            case_type = CaseTypeXpath(USERCASE_TYPE).case()
+            case = UserCaseXPath(case_type).case()
```
Pretty sure this should just be `case = UserCaseXPath(USERCASE_TYPE).case()`

kaapstorm committed:
Drop unnecessary changes

kaapstorm committed:
README outdated and unnecessary

kaapstorm committed:
Call Center reference slipped through the cracks

gcapalbo commented:
@kaapstorm Regarding the questions on CriticalSection -

General case and form processing uses some different locking mechanisms, and I'm not sure about the call center code as I haven't worked with it.  But basically if you ever notice situations where you're getting doc update conflicts or weird race condition issues, CriticalSection is a lightweight way to fix it.

The locking you implemented looks good - we need to use CriticalSection in enable_usercase in the off chance that it gets enabled simultaneously in two places (to prevent a doc update conflict on the domain).

Regarding your question on sync_user_case - if someone enables usercase and then right afterwards updates a user, it looks like the net effect is that sync_user_case will just be called twice for that user, right?  And since sync_user_case uses the CriticalSection, it will always make sure that the syncs happen serially and never simultaneously.  So I think it's good - the lock on enable_usercase makes sure that the domain doesn't have a doc update conflict, and the lock on sync_user_case makes sure that the sync process doesn't have a doc update conflict.

gcapalbo commented on corehq/apps/app_manager/tasks.py:
```diff
@@ -0,0 +1,9 @@
+from corehq.apps.users.models import CommCareUser
+
+
+@task
```
Do you have a sense of what the priority is of this task?  By default this task will run on the main celery queue which is the high priority queue.  If you think this is a longer running / lower priority task then we could move it to the background_queue (see https://github.com/dimagi/commcare-hq/blob/9e7b34ca0cded7d8da8c89c1f081c7d5460741f6/corehq/apps/users/tasks.py#L31 for an example).

No need to block on this, just might be worth giving some thought.

snopoke replied:
+1

gcapalbo commented on corehq/apps/app_manager/util.py:
```diff
@@ -359,3 +380,31 @@ def get_commcare_versions(request_user):
+
+
+def enable_usercase(domain_name):
+    with CriticalSection([domain_name + '_enable_usercase']):
```
No need to block on this, but I've been using the convention of 'keybase_%s' % identifier for the key names (i.e., key base first, then unique object identifier) to make it harder to have key name conflicts.

kaapstorm committed:
Add usercase datums, and usercase ID to session

snopoke commented on corehq/apps/app_manager/suite_xml.py:
```diff
@@ -1490,6 +1496,34 @@ def get_auto_select_assertions(self, case_id_xpath, mode, locale_arguments=None)
+        datums = []
+        if form.form_type == 'module_form' and uses_usercase(form):
+            if not self.is_usercase_enabled:
+                raise SuiteError('Form uses usercase, but usercase not enabled')
```
Does this error appear to the user? If so should be translated.

kaapstorm committed:
Use background queue for creating user cases

kaapstorm committed:
Keep CriticalSection key convention

snopoke commented on corehq/apps/app_manager/util.py:
```diff
@@ -359,3 +381,31 @@ def get_commcare_versions(request_user):
+
+def actions_use_usercase(actions):
+    return (any_usercase_items(actions['update_case']['update'].iterkeys()) or
+            any_usercase_items(actions['case_preload']['preload'].itervalues()))
```
could make this `action.get('update_case', {}).get('update', {}).iterkeys()` which would make it useful in the SuiteGenerator with `form.active_actions()`

snopoke commented on corehq/apps/app_manager/const.py:
```diff
@@ -28,5 +28,8 @@
 ATTACHMENT_PREFIX = 'attachment:'
 
 CASE_ID = 'case_id'
+USERCASE_TYPE = 'commcare-user'
```
Need to make sure user's can't create their own case type with this name.

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -976,6 +985,44 @@ def already_has_meta(self):
+    def add_usercase(self, form):
+        from corehq.apps.app_manager.util import split_path, get_usercase_keys, get_usercase_values
+
+        usercase_path = 'usercase/'
```
Maybe prefix with `commcare_`

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -1422,6 +1472,9 @@ def check_case_type(action):
                 for nodeset, property in action.preload.items():
+                    if property.startswith(USERCASE_PREFIX):
+                        # Ignore usercase properties
+                        continue
```
Don't need this here for AdvancedForms

snopoke commented on corehq/apps/app_manager/xform.py:
```diff
@@ -976,6 +985,44 @@ def already_has_meta(self):
+                self.add_setvalue(
+                    ref=nodeset,
+                    value=id_xpath.case().property(property_xpath),
+                )
```
Considering the similarity to normal preload block it should be easy to extract this to a function.

snopoke commented on corehq/apps/smsforms/app.py:
```diff
@@ -56,9 +61,20 @@ def start_session(domain, contact, app, module, form, case_id=None, yield_respon
+            if contact.doc_type == 'CommCareUser' and suite_gen.any_usercase_datums(extra_datums):
+                usercase = get_case_by_domain_hq_user_id(domain, contact.get_id, include_docs=False)
+                if usercase:
+                    session_data[USERCASE_ID] = usercase['id']
```
Probably worth extracting this to a util function `get_cloudcare_session_data`

snopoke commented on corehq/apps/smsforms/app.py:
```diff
@@ -56,9 +61,20 @@ def start_session(domain, contact, app, module, form, case_id=None, yield_respon
+            _assert = soft_assert(['nhooper_at_dimagi_dot_com'.replace('_at_', '@').replace('_dot_', '.')])
+            _assert(False, 'form uses usercase, but app_manager did not enable usercase for the domain')
+        else:
+            datums.extend(extra_datums)
```
I don't think this is necessary since `datums` isn't used below.

kaapstorm committed:
Merge branch 'master' into usercase

gcapalbo commented on corehq/apps/smsforms/app.py:
```diff
@@ -56,9 +61,20 @@ def start_session(domain, contact, app, module, form, case_id=None, yield_respon
         }
     
     if app and form:
```
Before we merge this we should make sure that the tests in the following files pass (they must be run manually by uncommenting them from the corresponding tests/__init__.py file):

corehq.apps.sms.tests.inbound_handlers
corehq.apps.kookoo.tests.outbound

Instructions for how to setup the test environment are here: https://github.com/dimagi/commcare-hq/blob/6b864322d159cb0f0d498eea57d234eeea6abd15/corehq/apps/sms/tests/util.py#L43

snopoke replied:
Thanks, good to know.

kaapstorm replied:
kookoo tests pass. corehq.apps.sms.tests.inbound_handlers.KeywordTestCase.test_all_inbound fails at line 124 with "AssertionError: u'arm_a' != 'arm_b'" in both this branch and on master. Checking it out.

gcapalbo replied:
@kaapstorm Sorry I should have mentioned sometimes the tests fail for some people because of something funny with couch where there needs to be a delay before querying for something after saving it. I ran all the manual sms and ivr tests and they passed, so no need to rerun it, but I'm guessing if you add sleep(1) (or some bigger number) in these places it would pass for you:

https://github.com/dimagi/commcare-hq/blob/9baeaa823a47063f958f30c380fa43ea3e0ad97a/corehq/apps/sms/tests/util.py#L240
https://github.com/dimagi/commcare-hq/blob/9baeaa823a47063f958f30c380fa43ea3e0ad97a/corehq/apps/sms/tests/util.py#L263

kaapstorm replied:
Thanks @gcapalbo, you're right. Adding sleep(1) fixed it. All passing now.

kaapstorm committed:
Reuse actions_use_usercase in SuiteGenerator

kaapstorm committed:
Separate get_cloudcare_session_data function

kaapstorm committed:
Separate add_case_preloads function

Changed util functions to accept dicts instead of lists of pairs

kaapstorm committed:
AdvancedForms don't need to skip usercase properties

kaapstorm committed:
Fixed data types

kaapstorm committed:
Merge branch 'master' into usercase

kaapstorm committed:
Add "User-As-A-Case" feature flag

kaapstorm committed:
Better check for usercase actions

More robust. Easier to read.

kaapstorm committed:
Only check "user-as-a-case" toggle if it's not already enabled.

If it's already enabled for the domain, don't bother checking the toggle, and don't trying to enable it.

kaapstorm committed:
Make usercase tag "commcare_usercase"

czue commented:
@kaapstorm let me know if you want me to buddy-review this beast, though at this point I'm tempted to let @snopoke and @gcapalbo take it over the finish line.

czue commented:
(although actually it's not quite as big as I thought so am open to reviewing if you think it'd be valuable)

kaapstorm commented:
Thanks for the offer @czue. @snopoke and @gcapalbo have given excellent feedback. You're welcome to join, but I think it'll be OK for you to skip this one.

kaapstorm committed:
Merge branch 'master' into usercase

snopoke commented:
Going to merge this now expecting that there are going to be some breaking changes in the next few days. e.g. changing `user:` to `user/`
